File: android/viewmodels/PhotoUploadViewModel.kt
package com.lookmaxxing.viewmodels

import android.content.Context
import android.net.Uri
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.lookmaxxing.utils.ImageCompressor
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException

/**
 * ViewModel that compresses a photo client-side, uploads it to the
 * lookmaxxing API, and polls for analysis completion.
 *
 * Usage (Compose):
 * ```
 * val vm: PhotoUploadViewModel = viewModel()
 * vm.upload(context, imageUri)
 * val state by vm.uploadState.collectAsState()
 * ```
 */
class PhotoUploadViewModel : ViewModel() {

    // ── UI state ────────────────────────────────────────────────
    data class UploadUiState(
        val isUploading: Boolean = false,
        val photoId: String? = null,
        val fileUrl: String? = null,
        val analysisStatus: String? = null, // "processing" | "completed" | "failed"
        val score: Double? = null,
        val error: String? = null,
        val progressMessage: String = ""
    )

    private val _state = MutableStateFlow(UploadUiState())
    val uploadState: StateFlow<UploadUiState> = _state

    // ── Configuration ───────────────────────────────────────────
    var apiBaseUrl: String = "https://lookmaxx-api.onrender.com/api/v1"
    var authToken: String = ""

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
        .writeTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
        .readTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
        .build()

    // ── Public API ──────────────────────────────────────────────

    /**
     * Compress + upload the selected image, then poll for status.
     */
    fun upload(context: Context, imageUri: Uri) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isUploading = true, error = null,
                progressMessage = "Compressing…")

            try {
                // 1. Compress
                val compressedData = withContext(Dispatchers.IO) {
                    ImageCompressor.compressImage(context, imageUri)
                } ?: throw IOException("Image compression returned null")

                val fileSizeKB = compressedData.size / 1024.0
                _state.value = _state.value.copy(
                    progressMessage = "Compressed to ${"%.1f".format(fileSizeKB)} KB"
                )

                // 2. Upload
                _state.value = _state.value.copy(progressMessage = "Uploading…")
                val photoResponse = withContext(Dispatchers.IO) {
                    uploadBytes(compressedData, "photo.jpg")
                }

                val photoId = photoResponse.getString("id")
                val status = photoResponse.optString("analysis_status", "processing")
                _state.value = _state.value.copy(
                    photoId = photoId,
                    fileUrl = photoResponse.optString("file_url"),
                    analysisStatus = status,
                    progressMessage = "Upload done — analysing…"
                )

                // 3. Poll for completion
                if (status == "processing" || status == "pending") {
                    pollStatus(photoId)
                }
            } catch (e: Exception) {
                _state.value = _state.value.copy(
                    error = e.message ?: "Unknown error",
                    isUploading = false
                )
            }
        }
    }

    /** Poll GET /photos/{id}/status every 1.5 s until completed or failed. */
    private suspend fun pollStatus(photoId: String) {
        for (attempt in 1..15) {
            delay(1_500)
            try {
                val json = withContext(Dispatchers.IO) { getStatus(photoId) }
                val status = json.optString("analysis_status", "processing")
                _state.value = _state.value.copy(analysisStatus = status)

                when (status) {
                    "completed" -> {
                        _state.value = _state.value.copy(
                            score = json.optDouble("score"),
                            isUploading = false,
                            progressMessage = "Complete!"
                        )
                        return
                    }
                    "failed" -> {
                        _state.value = _state.value.copy(
                            error = "Analysis failed on server",
                            isUploading = false
                        )
                        return
                    }
                }
            } catch (_: Exception) {
                // Retry on network hiccup
            }
        }
        _state.value = _state.value.copy(
            error = "Analysis timed out after 15 polls",
            isUploading = false
        )
    }

    // ── HTTP helpers ────────────────────────────────────────────

    @Throws(IOException::class)
    private fun uploadBytes(data: ByteArray, fileName: String): JSONObject {
        val mediaType = "image/jpeg".toMediaType()
        val body = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart(
                "file", fileName,
                data.toRequestBody(mediaType)
            )
            .build()

        val request = Request.Builder()
            .url("$apiBaseUrl/photos/upload")
            .header("Authorization", "Bearer $authToken")
            .post(body)
            .build()

        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                val errBody = response.body?.string() ?: ""
                throw IOException("Upload failed (HTTP ${response.code}): $errBody")
            }
            val responseBody = response.body?.string() ?: throw IOException("Empty response body")
            return JSONObject(responseBody)
        }
    }

    @Throws(IOException::class)
    private fun getStatus(photoId: String): JSONObject {
        val request = Request.Builder()
            .url("$apiBaseUrl/photos/$photoId/status")
            .header("Authorization", "Bearer $authToken")
            .get()
            .build()

        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw IOException("Status check failed (HTTP ${response.code})")
            }
            val body = response.body?.string() ?: throw IOException("Empty response body")
            return JSONObject(body)
        }
    }
}

File: ios/ViewModels/PhotoUploadViewModel.swift
import Foundation
import UIKit

/// ViewModel that handles photo selection, compression, and upload.
///
/// Usage:
/// ```
/// let vm = PhotoUploadViewModel()
/// vm.upload(selectedImage) { result in
///     switch result {
///     case .success(let response):  print("Photo ID: \(response.id)")
///     case .failure(let err):       print("Upload failed: \(err)")
///     }
/// }
/// ```
@MainActor
final class PhotoUploadViewModel: ObservableObject {
    @Published var isUploading = false
    @Published var uploadProgress: Float = 0.0  // 0 → 1

    private let apiBaseURL: String
    private var authToken: String

    init(apiBaseURL: String = "https://lookmaxx-api.onrender.com/api/v1",
         authToken: String = "") {
        self.apiBaseURL = apiBaseURL
        self.authToken = authToken
    }

    // ── Public API ──────────────────────────────────────────────

    /// Compress + upload a single photo.  Compression runs on a background
    /// queue; the upload call is async/await.
    func upload(_ image: UIImage) async throws -> PhotoUploadResponse {
        isUploading = true
        uploadProgress = 0.0
        defer { isUploading = false }

        // 1. Compress
        guard let compressedData = ImageCompressor.compressImage(image) else {
            throw UploadError.compressionFailed
        }
        let fileSizeKB = Double(compressedData.count) / 1024.0
        print("📦 Compressed to \(String(format: "%.1f", fileSizeKB)) KB")
        uploadProgress = 0.3

        // 2. Build multipart request
        let url = URL(string: "\(apiBaseURL)/photos/upload")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")

        let boundary = UUID().uuidString
        request.setValue("multipart/form-data; boundary=\(boundary)",
                         forHTTPHeaderField: "Content-Type")

        let body = multipartBody(data: compressedData,
                                 fileName: "photo.jpg",
                                 fieldName: "file",
                                 boundary: boundary)
        request.httpBody = body
        uploadProgress = 0.5

        // 3. Upload
        let (responseData, urlResponse) = try await URLSession.shared.data(for: request)

        guard let httpResponse = urlResponse as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw UploadError.serverError(code: (urlResponse as? HTTPURLResponse)?.statusCode ?? 0)
        }

        uploadProgress = 1.0

        // 4. Decode
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return try decoder.decode(PhotoUploadResponse.self, from: responseData)
    }

    /// Poll for analysis status until completed or failed.
    func pollStatus(photoID: String, maxAttempts: Int = 15, interval: TimeInterval = 1.5) async throws -> APIService.PhotoStatusResponse {
        let url = URL(string: "\(apiBaseURL)/photos/\(photoID)/status")!
        var request = URLRequest(url: url)
        request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")

        for _ in 0..<maxAttempts {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let httpResponse = response as? HTTPURLResponse,
                  (200...299).contains(httpResponse.statusCode) else {
                throw UploadError.serverError(code: (response as? HTTPURLResponse)?.statusCode ?? 0)
            }
            let decoder = JSONDecoder()
            let status = try decoder.decode(APIService.PhotoStatusResponse.self, from: data)
            switch status.analysis_status {
            case "completed", "failed":
                return status
            default:
                try await Task.sleep(nanoseconds: UInt64(interval * 1_000_000_000))
            }
        }
        throw UploadError.pollingTimeout
    }

    // MARK: - Private helpers

    private func multipartBody(data: Data,
                               fileName: String,
                               fieldName: String,
                               boundary: String) -> Data {
        var body = Data()
        let lineBreak = "\r\n".data(using: .utf8)!

        body.append("--\(boundary)\(lineBreak)".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"\(fieldName)\"; filename=\"\(fileName)\"\(lineBreak)".data(using: .utf8)!)
        body.append("Content-Type: image/jpeg\(lineBreak)\(lineBreak)".data(using: .utf8)!)
        body.append(data)
        body.append("\(lineBreak)--\(boundary)--\(lineBreak)".data(using: .utf8)!)

        return body
    }

    // MARK: - Error

    enum UploadError: LocalizedError {
        case compressionFailed
        case serverError(code: Int)
        case pollingTimeout

        var errorDescription: String? {
            switch self {
            case .compressionFailed:  return "Image compression failed"
            case .serverError(let c): return "Server error (HTTP \(c))"
            case .pollingTimeout:     return "Analysis timed out"
            }
        }
    }
}

